autobahn/wamp_tcp_client.hpp

```diff
@@ -0,0 +1,111 @@
+
+#ifndef AUTOBAHN_TCP_CLIENT_H
+#define AUTOBAHN_TCP_CLIENT_H
+
+#include <autobahn/autobahn.hpp>
+#include <boost/asio.hpp>
+#include <boost/version.hpp>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <tuple>
+namespace autobahn {
+/**
+ * @brief a class that wraps initialization of a raw tcp socket and WAMP client session utilizing it
+ */
+    class wamp_tcp_client {
+        typedef autobahn::wamp_session<boost::asio::ip::tcp::socket, boost::asio::ip::tcp::socket> wamp_tcp_session_t;
+    public:
+        wamp_tcp_client(std::shared_ptr<boost::asio::io_service> pIo,
+                        const boost::asio::ip::tcp::endpoint &Endpoint,
+                        const std::string &Realm,
+                        bool Debug = false) :
+                m_realm(Realm),
+                m_rawsocket_endpoint(Endpoint),
+                m_pIo(pIo),
+                m_debug(Debug) {
+            init();
+        }
+
+        wamp_tcp_client(std::shared_ptr<boost::asio::io_service> pIo,
+                        const std::string &IpAddress,
+                        uint16_t Port,
+                        const std::string &Realm,
+                        bool Debug = false) :
+                m_realm(Realm),
+                m_rawsocket_endpoint(boost::asio::ip::address::from_string(IpAddress), Port),
+                m_pIo(pIo),
+                m_debug(Debug) {
+            init();
+        }
+
+        /**
+     * @brief initialize members, creating shared pointers for the socket, and session
+     */
+        void init() {
+            m_pSession.reset();
+            m_pSocket.reset();
+            m_connected = boost::promise<bool>();
+            m_pSocket = std::make_shared<boost::asio::ip::tcp::socket>(*m_pIo);
+            m_pSession = std::make_shared<wamp_tcp_session_t>(*m_pIo, *m_pSocket, *m_pSocket, m_debug);
+        }
+
+        /**
+     * @brief launches the session asynchronously, returns a future containing an error code (0 means success)
+     */
+        boost::future<bool> launch() {
+            m_pSocket->async_connect(m_rawsocket_endpoint, [&](boost::system::error_code ec) {
+                if (!ec) {
+                    m_startFuture = m_pSession->start().then([&](boost::future<bool> started) {
+                        m_joinFuture = m_pSession->join(m_realm).then([&](boost::future<uint64_t> connected) {
+                            m_connected.set_value(true);
+                        });
+                    });
+                } else {
+                    std::cerr << "connect failed: " << ec.message() << std::endl;
+                    m_connected.set_value(false);
+                }
+            });
+            return m_connected.get_future();
+        }
+
+        ~wamp_tcp_client() {
+        }
+
+        /**
+     * This operator is used to pass through calls to the WAMP session. i.e *(pClient)->call(...)
+     */
+        std::shared_ptr<wamp_tcp_session_t> &operator->() {
```

Comment:
Since `operator->` is never meant to copy the actual session around, could it just return a regular pointer (`m_session.get()`) instead of the `shared_ptr`? Then unnecessary refcounting (one up, one down) can be avoided. Also AutobahnCpp sticks the `&` and `*` in function signatures to the type, rather than to the function or parameter name (this is something I overlooked in my list above and also files under "consistency with existing files"). All in all, I think it should be:
```
    wamp_tcp_session_t* operator->()
    {
        return m_session.get();
    }
```
Maybe a const version of `operator->` is also a good idea?

Lastly, I believe `(*client)->call(...)` would make for clearer code, because that makes it clear that it first dereferences a client pointer, and then still calls `operator->` on it. With the star on the outside, you have to rely on the greedy parsing behaviour of C++ (== not as clear), and the parentheses don't make much of a difference.